parastell/invessel_build.py

```diff
@@ -501,32 +508,325 @@ def extract_solids_and_mat_tags(self):
+        surface_keys = list(self.Surfaces.keys())
+
+        for i, key in enumerate(surface_keys[:-1]):
+            # Inner surface of given component = outer surface of inner
+            # component
+            if surface_keys[i + 1] == component:
+                surfaces.append(self.Surfaces[key])
+        surfaces.append(self.Surfaces[component])
+
+        coords = []
+        for surface in surfaces:
+            for rib in surface.Ribs:
+                coords.extend(rib.rib_loci)
+        coords = np.array(coords)
+
+        self.verts = self.mesh_mbc.create_vertices(coords)
+
+        self.mesh_set = self.mesh_mbc.create_meshset()
+        self.mesh_mbc.add_entity(self.mesh_set, self.verts)
+
+        for toroidal_idx in range(self._num_ribs - 1):
+            for poloidal_idx in range(self._num_rib_pts - 1):
+                self._create_tets_from_hex(poloidal_idx, toroidal_idx)
+
+    def _create_tets_from_hex(self, poloidal_idx, toroidal_idx):
+        """Creates five tetrahedra from defined hexahedron.
+        (Internal function not intended to be called externally)
+
+        Arguments:
+            poloidal_idx (int): index defining location along poloidal angle axis.
+            toroidal_idx (int): index defining location along toroidal angle axis.
+        """
+        # Relative offsets of vertices in a 3-D index space
+        hex_vertex_stencil = np.array(
+            [
+                [0, 0, 0],
+                [1, 0, 0],
+                [1, 1, 0],
+                [0, 1, 0],
+                [0, 0, 1],
+                [1, 0, 1],
+                [1, 1, 1],
+                [0, 1, 1],
+            ]
+        )
+
+        # IDs of hex vertices applying offset stencil to current point
+        hex_idx_data = (
+            np.array([0, poloidal_idx, toroidal_idx]) + hex_vertex_stencil
+        )
+
+        idx_list = [
+            self._get_vertex_id(vertex_idx) for vertex_idx in hex_idx_data
+        ]
+
+        # Define MOAB canonical ordering of hexahedron vertex indices
+        # Ordering follows right hand rule such that the fingers curl around
+        # one side of the tetrahedron and the thumb points to the remaining
+        # vertex. The vertices are ordered such that those on a face are first,
+        # ordered clockwise relative to the thumb, followed by the remaining
+        # vertex at the end of the thumb.
+        # See Moreno, Bader, Wilson 2024 for hexahedron splitting
+        hex_to_tets_mapping = [
+            [idx_list[0], idx_list[3], idx_list[1], idx_list[4]],
+            [idx_list[1], idx_list[3], idx_list[2], idx_list[6]],
+            [idx_list[1], idx_list[4], idx_list[6], idx_list[5]],
+            [idx_list[3], idx_list[6], idx_list[4], idx_list[7]],
+            [idx_list[1], idx_list[3], idx_list[6], idx_list[4]],
+        ]
+
+        for vertex_ids in hex_to_tets_mapping:
+            self._create_tet(vertex_ids)
+
+    def _get_vertex_id(self, vertex_idx):
```

Comment:
we should reuse code from source mesh code

Reply:
I'm sure we can find a way to make this method compatible with both method mesh routines.